diff --git a/.gitignore b/.gitignore
index b4ea3c6..22bbcee 100644
--- a/.gitignore
+++ b/.gitignore
@@ -72,3 +72,13 @@ Thumbs.db
 *.exe
 
 CMakeLists.txt.user
+*.deb
+*.changes
+*.buildinfo
+*.dsc
+obj-*
+debian/*.substvars
+debian/.debhelper/
+debian/debhelper-build-stamp
+debian/dir-view-test/
+debian/files
diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
new file mode 100644
index 0000000..d13a955
--- /dev/null
+++ b/.gitlab-ci.yml
@@ -0,0 +1,104 @@
+stages:
+  - build
+  - package
+  - package-test
+
+image: ${ASTRA_IMAGE}
+
+variables:
+    BUILD_DIR: build
+    ASTRA_IMAGE: registry.gitflic.ru/project/pilotdev/astra-linux/astra-se:1.8
+    DEB_PACKAGE: dir-view-test_1.0-1_amd64.deb 
+    PACKAGE_NAME: dir-view-test
+
+before_script:
+    - apt-get update
+    - apt-get install -y apt-transport-https ca-certificates
+    - echo "deb https://dl.astralinux.ru/astra/frozen/1.8_x86-64/1.8.1/uu/1/main-repository stable main" > /etc/apt/sources.list.d/astra-uu-main.list
+    - echo "deb https://dl.astralinux.ru/astra/frozen/1.8_x86-64/1.8.1/uu/1/extended-repository stable main" > /etc/apt/sources.list.d/astra-uu-extended.list
+    - apt-cache search build-essential
+    - apt-cache search cmake
+    - apt-get update
+    - apt-get install -y build-essential git cmake qtbase5-dev qt5-default qtdeclarative5-dev qtquickcontrols2-5-dev qttools5-dev
+    - apt-get install -y debhelper
+
+build-job:
+  stage: build
+  tags: 
+    - docker
+  script:
+    - ls -la
+    - cmake -S . -B ${BUILD_DIR}
+    - cmake --build ${BUILD_DIR}
+  artifacts:
+    name: ${CI_PROJECT_NAME}-${CI_COMMIT_REF_NAME}-${CI_COMMIT_SHORT_SHA}-build
+    when: always
+    expire_in: 1 week
+    paths:
+      - ${BUILD_DIR}/dir_view_test
+      - ${BUILD_DIR}/Makefile
+      - ${BUILD_DIR}/CMakeCache.txt
+      - ${BUILD_DIR}/CPackSourceConfig.cmake
+      - ${BUILD_DIR}/CPackConfig.cmake
+
+
+dpkg-buildpackage-job:
+  stage: package
+  tags:
+    - docker
+  script:
+    - ls -la
+    - ls -la ..
+    - rm -rf build
+    - git archive --format=tar e31005e | xz > ../dir-view-test_1.0.orig.tar.xz
+    - dpkg-buildpackage -us -uc
+    - mkdir package
+    - mv ../*.deb package/
+    - ls -lR ..
+  artifacts:
+    name: ${CI_PROJECT_NAME}-${CI_COMMIT_REF_NAME}-${CI_COMMIT_SHORT_SHA}-dpkg-buildpackage
+    when: always
+    expire_in: 1 week
+    paths:
+      - package/
+  needs:
+    - build-job
+
+cpack-job:
+  stage: package
+  tags: 
+    - docker
+  script:
+    - cd ${BUILD_DIR}
+    - cpack -G DEB
+    - mkdir -p ../package
+    - mv *.deb ../package
+    - ls -lR ..
+  artifacts:
+    name: ${CI_PROJECT_NAME}-${CI_COMMIT_REF_NAME}-${CI_COMMIT_SHORT_SHA}-cpack
+    when: always
+    expire_in: 1 week
+    paths:
+      - package/
+  needs:
+    - build-job
+
+
+package-test-job:
+  stage: package-test
+  tags: 
+    - docker
+  script:
+    - ls -la
+    - find . -name "*.deb"
+    - echo "[INFO] Package installation"
+    - dpkg -i package/${DEB_PACKAGE} || true
+    - apt install -f -y
+    - echo "[INFO] Check files in the debian package"
+    - dpkg -c ${DEB_PACKAGE}
+    - which dir_view_test
+    - echo "[INFO] Check installed files"
+    - dpkg -L ${PACKAGE_NAME}
+  needs: 
+    - dpkg-buildpackage-job
+    - cpack-job
diff --git a/CMakeLists.txt b/CMakeLists.txt
index f10e735..2ce479b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,6 +19,12 @@ set(PROJECT_SOURCES
         mainwindow.cpp
         mainwindow.h
         mainwindow.ui
+        dir_tree_model.h
+        dir_tree_model.cpp
+        dir_tree_view.h
+        dir_tree_view.cpp
+        dir_filter_model.h
+        dir_filter_model.cpp
 )
 
 if(${QT_VERSION_MAJOR} GREATER_EQUAL 6)
@@ -57,3 +63,17 @@ set_target_properties(dir_view_test PROPERTIES
 if(QT_VERSION_MAJOR EQUAL 6)
     qt_finalize_executable(dir_view_test)
 endif()
+
+install(TARGETS dir_view_test RUNTIME DESTINATION bin)
+
+set(CPACK_PACKAGE_NAME "dir_tree_view")
+set(CPACK_PACKAGE_VERSION "1.0")
+set(CPACK_PACKAGE_CONTACT "sokol28.nik@gmail.com")
+
+set(CPACK_DEBIAN_PACKAGE_NAME "dir_view_test")
+set(CPACK_DEBIAN_PACKAGE_MAINTAINER "sokol")
+set(CPACK_DEBIAN_PACKAGE_CONTACT "sokol28.nik@gmail.com")
+set(CPACK_DEBIAN_PACKAGE_DESCRIPTION "A graphical application for viewing the file system tree with support for filtering, sorting, and searching")
+
+set(CPACK_GENERATOR "DEB")
+include(CPack)
diff --git a/LICENSE b/LICENSE
new file mode 100644
index 0000000..7922633
--- /dev/null
+++ b/LICENSE
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2026 eyevievv
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
diff --git a/README.md b/README.md
new file mode 100644
index 0000000..6907fc0
--- /dev/null
+++ b/README.md
@@ -0,0 +1,154 @@
+# Dir Tree Viewer
+
+Графическое приложение на Qt для просмотра дерева файловой системы с поддержкой фильтрации, сортировки и поиска.
+
+## Описание
+
+Приложение отображает структуру файловой системы в виде дерева и позволяет выполнять поиск файлов и каталогов по имени. 
+По умолчанию в качестве стартовой директории используется домашний каталог текущего пользователя. 
+
+Проект разработан на основе [dirview](https://doc.qt.io/archives/qt-5.15/qtwidgets-itemviews-dirview-example.html).
+
+<img width="807" height="634" alt="image" src="https://github.com/user-attachments/assets/fa639cff-874a-4060-b8ca-b2dad2214fbc" />
+
+## Возможности
+### Фильтрация
+
+Пользователь вводит в строку поиска регулярное выражение, по которому ищутся файлы и директории. В обычном режиме поиск идет только в корневой директории дерева. 
+В режиме глубокого поиска рекурсивно обходятся все подкаталоги. В дереве отображаются найденные элементы, а также директории,
+необходимые для сохранения пути к найденным элементам.
+
+<img width="808" height="634" alt="image" src="https://github.com/user-attachments/assets/d3ebef48-1d0b-4597-aa8e-7e5511a9c68d" />
+
+Также доступна функция учета регистра.
+
+### Сортировка
+
+При нажатии на заголовок столбца выполняется сортировка элементов дерева
+по соответствующему столбцу в порядке возрастания или убывания.
+
+### Размер директорий
+
+Для каталогов отображается надпись "Обновить", при двойном щелчке на которую выполняется рекурсивный подсчёт размера каталога.
+Полученный результат сохраняется в кэше и отображается в таблице.
+
+## Структура проекта
+
+```text
+├── CMakeLists.txt
+├── dir_filter_model.cpp
+├── dir_filter_model.h
+├── dir_tree_model.cpp
+├── dir_tree_model.h
+├── dir_tree_view.cpp
+├── dir_tree_view.h
+├── main.cpp
+├── mainwindow.cpp
+├── mainwindow.h
+├── mainwindow.ui
+├── TASK.md
+└── README.md
+```
+
+## Компоненты приложения
+
+### DirTreeModel
+
+Наследует `QFileSystemModel`. Представляет собой модель системы файлов - предоставляет информацию о файлах и директориях файловой системы. 
+
+Отвечает за работу с файловой системой и хранение кэша рассчитанных размеров. 
+
+Поддерживает ленивую загрузку по мере обращения к узлам дерева.
+
+### DirFilterModel
+
+Наследует `QSortFilterProxyModel`. Является прокси-посредником между моделью и отображением. 
+Применяет фильтрацию файлов по регулярному выражению. 
+
+Узел проходит фильтр, если выполняется хотя бы одно из условий:
+- имя узла содержит паттерн поиска.
+- один из потомков узла содержит паттерн (в режиме глубокого поиска).
+
+## DirTreeView
+
+Наследует QTreeView. Отвечает за визуальное отображение дерева. Принимает DirFilterModel как модель.
+
+### MainWindow
+
+Главное окно приложения.
+
+Помимо виджета дерева файлов содержит:
+- Строку поиска;
+- Кнопку «Учитывать регистр»;
+- Кнопку «Глубокий поиск».
+
+## Сборка
+
+### Требования
+
+- C++17+;
+- CMake;
+- Qt 5.15.
+
+### Сборка с CMake
+
+```bash
+cmake -S . -B build
+cmake --build build
+```
+
+### Запуск
+
+```bash
+./dir_view_test [опции] [директория]
+```
+
+### Параметры
+```text
+  -h, --help     Вывод справки по параметрам командной строки.
+  --help-all     Вывод справки с особыми параметрами Qt.
+  -v, --version  Показать сведения о версии.
+  -c             Отключить отображение пользовательских иконок.
+  -w             Отключить отслеживание изменений.
+```
+### Аргументы
+`директория` - начальная директория дерева (абсолютный путь)
+
+
+## Используемые технологии
+- C++
+- Qt
+  - Qt Widgets
+  - QTreeView
+  - QFileSystemModel
+  - QSortFilterProxyModel
+- CMake
+- git
+- dpkg
+- dh_make
+
+### Сборка пакета deb
+
+Для проекта подготовлена конфигурация Debian-пакета в каталоге `debian/`, включающая информацию о приложении.
+
+Изменения относительно исходного проекта dirview оформлены в виде патча `dir-view-test.patch` в `debian/patches/`.
+
+Сборка выполняется командой:
+
+```bash
+dpkg-buildpackage -us -uc
+```
+
+После успешной сборки в родительском каталоге будут созданы файлы:
+
+```bash
+../dir-view-test_1.0-1_amd64.deb
+../dir-view-test_1.0-1_amd64.buildinfo
+../dir-view-test_1.0-1_amd64.changes
+```
+
+Установить пакет можно командой:
+
+```bash
+sudo dpkg -i ../dir-view-test_1.0-1_amd64.deb
+```
diff --git a/dir_filter_model.cpp b/dir_filter_model.cpp
new file mode 100644
index 0000000..4f552b9
--- /dev/null
+++ b/dir_filter_model.cpp
@@ -0,0 +1,61 @@
+#include "dir_filter_model.h"
+#include <QString>
+#include <QDir>
+#include <QFileSystemModel>
+
+DirFilterModel::DirFilterModel(QObject *parent)
+    : QSortFilterProxyModel{parent}
+{
+
+}
+
+bool DirFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const
+{
+    QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);
+    if (!index.isValid()) return false;
+
+    auto fsModel = qobject_cast<QFileSystemModel*>(sourceModel());
+
+    if (fsModel->filePath(index) == rootPath_)
+        return true;
+
+    if (!currentSearchPattern_.isEmpty() && matchedPaths_.isEmpty()) {
+        return false;
+    }
+    else if (currentSearchPattern_.isEmpty()) {
+        return true;
+    }
+
+    QString name = sourceModel()->data(index).toString();
+    if (name == "." || name == "..") return false;
+
+    QString path = QDir::cleanPath(fsModel->filePath(index));
+
+    return matchedPaths_.contains(path);
+}
+
+void DirFilterModel::setRootPath(const QString& path)
+{
+    rootPath_ = QDir::cleanPath(path);
+}
+
+void DirFilterModel::setMatchedPaths(const QSet<QString> paths)
+{
+    matchedPaths_ = paths;
+    invalidateFilter();
+}
+
+void DirFilterModel::setSearchDepth(SearchDepth depth)
+{
+    searchDepth_ = depth;
+}
+
+void DirFilterModel::setSearchText(const QString& text)
+{
+    currentSearchPattern_ = text;
+}
+
+SearchDepth DirFilterModel::getSearchDepth() const
+{
+    return searchDepth_;
+}
diff --git a/dir_filter_model.h b/dir_filter_model.h
new file mode 100644
index 0000000..0aa08ea
--- /dev/null
+++ b/dir_filter_model.h
@@ -0,0 +1,35 @@
+#ifndef DIRFILTERMODEL_H
+#define DIRFILTERMODEL_H
+
+#include <QSortFilterProxyModel>
+#include <QWidget>
+#include <QSet>
+
+enum class SearchDepth { Shallow, Deep };
+
+class DirFilterModel : public QSortFilterProxyModel
+{
+    Q_OBJECT
+
+public:
+    explicit DirFilterModel(QObject *parent = nullptr);
+
+    void setRootPath(const QString& path);
+    void setSearchDepth(SearchDepth depth);
+    void setMatchedPaths(const QSet<QString> paths);
+    void setSearchText(const QString& text);
+
+    SearchDepth getSearchDepth() const;
+
+protected:
+    bool filterAcceptsRow(int sourceRow,
+                const QModelIndex &sourceParent) const override;
+
+
+    QSet<QString> matchedPaths_;
+    QString rootPath_;
+    SearchDepth searchDepth_ = SearchDepth::Shallow;
+    QString currentSearchPattern_;
+};
+
+#endif // DIRFILTERMODEL_H
diff --git a/dir_tree_model.cpp b/dir_tree_model.cpp
new file mode 100644
index 0000000..fab829b
--- /dev/null
+++ b/dir_tree_model.cpp
@@ -0,0 +1,51 @@
+#include "dir_tree_model.h"
+#include <QLocale>
+
+DirTreeModel::DirTreeModel(QObject *parent)
+    : QFileSystemModel(parent)
+{
+    this->setRootPath("");
+    this->setFilter(QDir::AllEntries | QDir::Hidden | QDir::System);
+}
+
+QVariant DirTreeModel::data(const QModelIndex &index, int role) const
+{
+    if (index.column() == 1 && role == Qt::DisplayRole) {
+        QFileInfo info = fileInfo(index);
+
+        if (info.isDir()) {
+            auto it = dirSizes_.find(info.absoluteFilePath());
+
+            if (it != dirSizes_.end())
+                return QLocale().formattedDataSize(it.value());
+
+            return "Обновить";
+        }
+    }
+    return QFileSystemModel::data(index, role);
+}
+
+void DirTreeModel::calculateDirSize(const QString &path)
+{
+    qint64 size = 0;
+
+    QDirIterator it(
+        path,
+        QDir::Files | QDir::Hidden | QDir::NoDotAndDotDot,
+        QDirIterator::Subdirectories
+    );
+
+    while (it.hasNext() && !it.fileInfo().isSymLink())
+    {
+        it.next();
+        size += it.fileInfo().size();
+    }
+
+    dirSizes_[path] = size;
+
+    QModelIndex idx = index(path);
+
+    emit dataChanged(this->index(idx.row(), 1, idx.parent()),
+                     this->index(idx.row(), 1, idx.parent()));
+}
+
diff --git a/dir_tree_model.h b/dir_tree_model.h
new file mode 100644
index 0000000..c93ffe4
--- /dev/null
+++ b/dir_tree_model.h
@@ -0,0 +1,20 @@
+#ifndef DIRTREEMODEL_H
+#define DIRTREEMODEL_H
+
+#include <QObject>
+#include <QFileSystemModel>
+#include <QHash>
+
+class DirTreeModel : public QFileSystemModel
+{
+public:
+    DirTreeModel(QObject *parent = nullptr);
+
+    QVariant data(const QModelIndex& index, int role = Qt::DisplayRole) const override;
+    void calculateDirSize(const QString& path);
+
+private:
+    QHash<QString, qint64> dirSizes_;
+};
+
+#endif // DIRTREEMODEL_H
diff --git a/dir_tree_view.cpp b/dir_tree_view.cpp
new file mode 100644
index 0000000..42478c9
--- /dev/null
+++ b/dir_tree_view.cpp
@@ -0,0 +1,18 @@
+#include "dir_tree_view.h"
+#include <QScroller>
+
+DirTreeView::DirTreeView(QWidget *parent)
+    : QTreeView(parent)
+{
+
+}
+
+void DirTreeView::setup()
+{
+    this->setAnimated(false);
+    this->setIndentation(20);
+
+    this->setSortingEnabled(true);
+
+    QScroller::grabGesture(this, QScroller::TouchGesture);
+}
diff --git a/dir_tree_view.h b/dir_tree_view.h
new file mode 100644
index 0000000..8f860fe
--- /dev/null
+++ b/dir_tree_view.h
@@ -0,0 +1,16 @@
+#ifndef DIRTREEVIEW_H
+#define DIRTREEVIEW_H
+
+#include <QWidget>
+#include <QTreeView>
+
+class DirTreeView : public QTreeView
+{
+public:
+    DirTreeView(QWidget *parent = nullptr);
+
+    void setup();
+
+};
+
+#endif // DIRTREEVIEW_H
diff --git a/main.cpp b/main.cpp
index 9e95bfa..d2688ad 100644
--- a/main.cpp
+++ b/main.cpp
@@ -1,70 +1,13 @@
 #include "mainwindow.h"
 
 #include <QApplication>
-#include <QFileSystemModel>
-#include <QFileIconProvider>
-#include <QScreen>
-#include <QScroller>
-#include <QTreeView>
-#include <QCommandLineParser>
-#include <QCommandLineOption>
 
 int main(int argc, char *argv[]) {
     QApplication app(argc, argv);
     QCoreApplication::setApplicationVersion(QT_VERSION_STR);
 
-    QCommandLineParser parser;
-    parser.setApplicationDescription("Утилита отображения дерева файловой системы");
-    parser.addHelpOption();
-    parser.addVersionOption();
-
-    QCommandLineOption dontUseCustomDirectoryIconsOption("c", "Отключить отображение пользовательских иконок.");
-    parser.addOption(dontUseCustomDirectoryIconsOption);
-
-    QCommandLineOption dontWatchOption("w", "Отключить отслеживание изменений.");
-    parser.addOption(dontWatchOption);
-
-    parser.addPositionalArgument("директория", "Начальная директория.");
-    parser.process(app);
-
-    const QString rootPath = parser.positionalArguments().isEmpty()
-        ? QString(QDir::homePath()) : parser.positionalArguments().first();
-
-
-    QFileSystemModel model;
-    model.setRootPath("");
-
-    if (parser.isSet(dontUseCustomDirectoryIconsOption)) {
-        model.setOption(QFileSystemModel::DontUseCustomDirectoryIcons);
-    }
-    if (parser.isSet(dontWatchOption)) {
-        model.setOption(QFileSystemModel::DontWatchForChanges);
-    }
-
-    QTreeView tree;
-    tree.setModel(&model);
-
-    if (!rootPath.isEmpty()) {
-        const QModelIndex rootIndex = model.index(QDir::cleanPath(rootPath));
-        if (rootIndex.isValid()) {
-            tree.setRootIndex(rootIndex);
-        }
-    }
-
-    // Demonstrating look and feel features
-//    tree.setAnimated(false);
-    tree.setIndentation(20);
-    tree.setSortingEnabled(true);
-
-    const QSize availableSize = tree.screen()->availableGeometry().size();
-    tree.resize(availableSize / 2);
-    tree.setColumnWidth(0, tree.width() / 3);
-
-    // Make it flickable on touchscreens
-    QScroller::grabGesture(&tree, QScroller::TouchGesture);
-
-    tree.setWindowTitle(QObject::tr("Dir View"));
-    tree.show();
+    MainWindow window(app);
+    window.show();
 
     return app.exec();
 }
diff --git a/mainwindow.cpp b/mainwindow.cpp
index a06fc8e..1f945ae 100644
--- a/mainwindow.cpp
+++ b/mainwindow.cpp
@@ -1,11 +1,83 @@
 #include "mainwindow.h"
 #include "./ui_mainwindow.h"
 
-MainWindow::MainWindow(QWidget *parent)
+#include <QDir>
+#include <QFileIconProvider>
+#include <QScreen>
+#include <QCommandLineOption>
+#include <QVBoxLayout>
+#include <QTimer>
+
+
+MainWindow::MainWindow(const QApplication& app, QWidget *parent)
     : QMainWindow(parent)
     , ui(new Ui::MainWindow)
 {
     ui->setupUi(this);
+
+    this->setWindowTitle("Dir Tree Viewer");
+
+    dirModel_ = new DirTreeModel(this);
+    dirView_ = new DirTreeView(this);
+    filterModel_ = new DirFilterModel(this);
+
+    lineEdit_ = new QLineEdit(this);
+    lineEdit_->setPlaceholderText("Поиск");
+    enableCaseSensetivityButton_ = new QCheckBox("Учитывать регистр", this);
+    enableDeepSearchButton_ = new QCheckBox("Глубокий поиск", this);
+
+    QCommandLineParser parser;
+    parseCommandLine(parser, app);
+
+    rootPath_ = parser.positionalArguments().isEmpty()
+        ? QString(QDir::homePath()) : parser.positionalArguments().first();
+
+    filterModel_->setRootPath(rootPath_);
+    filterModel_->setSourceModel(dirModel_);
+    filterModel_->setFilterCaseSensitivity(Qt::CaseInsensitive);
+    filterModel_->setRecursiveFilteringEnabled(true);
+
+    dirView_->setModel(filterModel_);
+
+
+
+    setupRootPath();
+
+    dirView_->setup();
+
+    setupLayout();
+
+    const QSize availableSize = \
+            dirView_->screen()->availableGeometry().size();
+    dirView_->resize(availableSize / 2);
+    dirView_->setColumnWidth(0, dirView_->width() / 3);
+
+    debounceTimer_ = new QTimer(this);
+    debounceTimer_->setSingleShot(true);
+
+
+    connect(this->lineEdit_, &QLineEdit::textChanged,
+            this, &MainWindow::pathChanged);
+    connect(this->enableCaseSensetivityButton_, &QCheckBox::clicked,
+            this, &MainWindow::enableCaseSensetivity);
+    connect(this->enableDeepSearchButton_, &QCheckBox::toggled,
+                this, &MainWindow::enableDeepSearch);
+
+    connect(this->dirModel_, &QFileSystemModel::directoryLoaded,
+            this, &MainWindow::debouncePath);
+    connect(this->dirView_, &QTreeView::doubleClicked,
+            this, &MainWindow::onTreeDoubleClicked);
+
+
+    connect(this->debounceTimer_, &QTimer::timeout, this, [this]() {
+        filterModel_->invalidate();
+    });
+
+    connect(this->lineEdit_, &QLineEdit::textChanged, this, [this](const QString& text) {
+        pendingText_ = text;
+        debounceTimer_->start(300);
+    });
+
 }
 
 MainWindow::~MainWindow()
@@ -13,3 +85,211 @@ MainWindow::~MainWindow()
     delete ui;
 }
 
+void MainWindow::parseCommandLine(QCommandLineParser& parser, const QApplication& app)
+{
+    parser.setApplicationDescription("Утилита отображения дерева файловой системы");
+    parser.addHelpOption();
+    parser.addVersionOption();
+
+    QCommandLineOption dontUseCustomDirectoryIconsOption("c", "Отключить отображение пользовательских иконок.");
+    parser.addOption(dontUseCustomDirectoryIconsOption);
+
+    QCommandLineOption dontWatchOption("w", "Отключить отслеживание изменений.");
+    parser.addOption(dontWatchOption);
+
+    parser.addPositionalArgument("директория", "Начальная директория.");
+
+    parser.process(app);
+
+    if (parser.isSet(dontUseCustomDirectoryIconsOption)) {
+        dirModel_->setOption(QFileSystemModel::DontUseCustomDirectoryIcons);
+    }
+    if (parser.isSet(dontWatchOption)) {
+        dirModel_->setOption(QFileSystemModel::DontWatchForChanges);
+    }
+
+}
+
+void MainWindow::setupLayout()
+{
+    QHBoxLayout* topLayout = new QHBoxLayout;
+
+    topLayout->addWidget(lineEdit_);
+    topLayout->addWidget(enableCaseSensetivityButton_);
+    topLayout->addWidget(enableDeepSearchButton_);
+
+
+    QWidget* central_widget = new QWidget(this);
+
+    QVBoxLayout* layout = new QVBoxLayout(central_widget);
+    layout->addLayout(topLayout);
+    layout->addWidget(dirView_);
+
+    setCentralWidget(central_widget);
+}
+
+void MainWindow::setupRootPath()
+{
+    if (!rootPath_.isEmpty()) {
+        const QModelIndex sourceIndex = dirModel_->index(QDir::cleanPath(rootPath_));
+        if (sourceIndex.isValid()) {
+            const QModelIndex proxyIndex = filterModel_->mapFromSource(sourceIndex);
+            dirView_->setRootIndex(proxyIndex);
+        }
+    }
+}
+
+void MainWindow::configureExpand(const QString& text)
+{
+    if (text.isEmpty() ||
+            filterModel_->getSearchDepth() == SearchDepth::Shallow
+    ) {
+        dirView_->collapseAll();
+    }
+    else {
+        expandMatching(dirView_->rootIndex());
+    }
+}
+
+void MainWindow::expandMatching(const QModelIndex &parent)
+{
+    const int rowsCount = filterModel_->rowCount(parent);
+    for (int i = 0; i < rowsCount; ++i) {
+        QModelIndex index = filterModel_->index(i, 0, parent);
+
+        if (filterModel_->rowCount(parent) > 0) {
+            dirView_->expand(index);
+            expandMatching(index);
+        }
+    }
+}
+
+QSet<QString> MainWindow::shallowSearchFindMatches(const QString &text)
+{
+    QSet<QString> result;
+
+    QDir dir(rootPath_);
+
+    QFileInfoList entries = dir.entryInfoList(
+        QDir::AllEntries | QDir::Hidden | QDir::NoDotAndDotDot);
+
+    for (const QFileInfo& info : entries) {
+        if (info.fileName().contains(text,
+            filterModel_->filterCaseSensitivity()))
+        {
+            addAncestors(info.absoluteFilePath(), result);
+        }
+    }
+
+    return result;
+}
+
+QSet<QString> MainWindow::deepSearchFindMatches(const QString &text)
+{
+    QSet<QString> result;
+    if (text.isEmpty()) return result;
+
+    QDirIterator it(
+        rootPath_,
+        QDir::AllEntries | QDir::Hidden | QDir::NoDotAndDotDot,
+        QDirIterator::Subdirectories
+    );
+
+    while (it.hasNext()) {
+        QString path = it.next();
+
+        QFileInfo info(path);
+
+        if (info.fileName().contains(text,
+            filterModel_->filterCaseSensitivity()))
+        {
+            addAncestors(path, result);
+        }
+    }
+
+    return result;
+}
+
+void MainWindow::addAncestors(QString path, QSet<QString> &result)
+{
+    path = QDir::cleanPath(path);
+
+    while (!path.isEmpty()) {
+        result.insert(path);
+
+        if (path == rootPath_) break;
+
+        QFileInfo info(path);
+        QString parentPath = info.dir().absolutePath();
+
+        if (parentPath == path) break;
+
+        path = parentPath;
+    }
+}
+
+void MainWindow::pathChanged(const QString& text)
+{
+    filterModel_->setSearchText(text);
+
+    if (text.isEmpty()) {
+        filterModel_->setMatchedPaths({});
+    }
+    else {
+        QSet<QString> matches;
+        if (filterModel_->getSearchDepth() == SearchDepth::Shallow) {
+            matches = shallowSearchFindMatches(text);
+        }
+        else {
+            matches = deepSearchFindMatches(text);
+        }
+
+        filterModel_->setMatchedPaths(matches);
+
+        QTimer::singleShot(0, this, [this]()
+        {
+            expandMatching(dirView_->rootIndex());
+        });
+    }
+
+    configureExpand(text);
+}
+
+void MainWindow::enableCaseSensetivity()
+{
+    if (filterModel_->filterCaseSensitivity() != Qt::CaseSensitive) {
+        filterModel_->setFilterCaseSensitivity(Qt::CaseSensitive);
+    }
+    else {
+        filterModel_->setFilterCaseSensitivity(Qt::CaseInsensitive);
+    }
+
+    const QString& text = lineEdit_->text();
+    pathChanged(text);
+}
+
+void MainWindow::enableDeepSearch(bool checked)
+{
+    filterModel_->setSearchDepth(static_cast<SearchDepth>(checked));
+    pathChanged(lineEdit_->text());
+}
+
+void MainWindow::debouncePath(const QString &path)
+{
+    if (!path.startsWith(rootPath_)) return;
+    debounceTimer_->start(150);
+}
+
+void MainWindow::onTreeDoubleClicked(const QModelIndex &proxyIndex)
+{
+    if (proxyIndex.column() != 1) return;
+
+    QModelIndex index = filterModel_->mapToSource(proxyIndex);
+    QFileInfo info = dirModel_->fileInfo(index);
+
+    if (!info.isDir()) return;
+
+    dirModel_->calculateDirSize(info.absoluteFilePath());
+}
+
+
diff --git a/mainwindow.h b/mainwindow.h
index 4643e32..32e96f3 100644
--- a/mainwindow.h
+++ b/mainwindow.h
@@ -1,7 +1,15 @@
 #ifndef MAINWINDOW_H
 #define MAINWINDOW_H
 
+#include "dir_tree_model.h"
+#include "dir_tree_view.h"
+#include "dir_filter_model.h"
+
 #include <QMainWindow>
+#include <QSortFilterProxyModel>
+#include <QCommandLineParser>
+#include <QLineEdit>
+#include <QCheckBox>
 
 QT_BEGIN_NAMESPACE
 namespace Ui { class MainWindow; }
@@ -12,10 +20,41 @@ class MainWindow : public QMainWindow
     Q_OBJECT
 
 public:
-    MainWindow(QWidget *parent = nullptr);
+    MainWindow(const QApplication& app, QWidget *parent = nullptr);
     ~MainWindow();
 
 private:
+    void parseCommandLine(QCommandLineParser& parser, const QApplication& app);
+    void setupRootPath();
+    void setupLayout();
+    void configureExpand(const QString& text);
+    void expandMatching(const QModelIndex& parent);
+    QSet<QString> shallowSearchFindMatches(const QString& text);
+    QSet<QString> deepSearchFindMatches(const QString& text);
+    void addAncestors(QString path, QSet<QString>& result);
+
+
     Ui::MainWindow *ui;
+    DirTreeModel* dirModel_;
+    DirTreeView* dirView_;
+
+    DirFilterModel* filterModel_;
+
+    QLineEdit* lineEdit_;
+    QCheckBox* enableCaseSensetivityButton_;
+    QCheckBox* enableDeepSearchButton_;
+
+    QString rootPath_;
+
+    QTimer* debounceTimer_;
+    QString pendingText_;
+
+private slots:
+    void pathChanged(const QString& text);
+    void enableCaseSensetivity();
+    void enableDeepSearch(bool checked);
+
+    void debouncePath(const QString& path);
+    void onTreeDoubleClicked(const QModelIndex& index);
 };
 #endif // MAINWINDOW_H
